dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs

```diff
@@ -0,0 +1,465 @@
+        {
+            result.FinalAnswer = finalAnswerMatch.Groups[1].Value.Trim();
+            return result;
+        }
+
+        // Extract thought
+        Match thoughtMatch = s_thoughtRegex.Match(input);
+
+        if (thoughtMatch.Success)
+        {
+            result.Thought = thoughtMatch.Value.Trim();
+        }
+        else if (!input.Contains(Action))
+        {
+            result.Thought = input;
+        }
+        else
+        {
+            throw new InvalidOperationException("Unexpected input format");
+        }
+
+        result.Thought = result.Thought.Replace(Thought, string.Empty).Trim();
+
+        // Extract action
+        Match actionMatch = s_actionRegex.Match(input);
+
+        if (actionMatch.Success)
+        {
+            var json = actionMatch.Groups[1].Value.Trim();
+
+            try
+            {
+                var systemStepResults = JsonSerializer.Deserialize<SystemStep>(json);
+
+                if (systemStepResults == null)
+                {
+                    result.Observation = $"System step parsing error, empty JSON: {json}";
+                }
+                else
+                {
+                    result.Action = systemStepResults.Action;
+                    result.ActionVariables = systemStepResults.ActionVariables;
+                }
+            }
+            catch (JsonException)
+            {
+                result.Observation = $"System step parsing error, invalid JSON: {json}";
+            }
+        }
+
+        return result;
+    }
+
+    private void AddExecutionStatsToContext(List<SystemStep> stepsTaken, SKContext context)
+    {
+        context.Variables.Set("stepCount", stepsTaken.Count.ToString(CultureInfo.InvariantCulture));
+        context.Variables.Set("stepsTaken", JsonSerializer.Serialize(stepsTaken));
+
+        // Materialize the nonEmptyActions collection into a List
+        var nonEmptyActions = stepsTaken.Where(s => !string.IsNullOrEmpty(s.Action)).ToList();
+        int skillCallCount = nonEmptyActions.Count;
+        string skillCallCountStr = skillCallCount.ToString(CultureInfo.InvariantCulture);
+
+        var distinctSkills = nonEmptyActions.Select(s => s.Action).Distinct().ToList();
+        string skillCallList = string.Join(", ", distinctSkills);
+
+        string skillCallListWithCounts = string.Join(", ", distinctSkills.Select(skill =>
+            $"{skill}({nonEmptyActions.Count(s => s.Action == skill)})"));
+
+        context.Variables.Set("skillCount", $"{skillCallCountStr} ({skillCallListWithCounts})");
+    }
+
+    private string CreateScratchPad(string question, List<SystemStep> stepsTaken)
+    {
+        if (stepsTaken.Count == 0)
+        {
+            return string.Empty;
+        }
+
+        var scratchPadLines = new List<string>();
+
+        // Add the original first thought
+        scratchPadLines.Add("This was your previous work (but I haven't seen any of it! I only see what you return as final answer):");
```

Comment:
Where does this string come from? Is it part of some mrkl implementation paper, or just made up? Just wondering, as it seems somewhat informal, and requires a concept of what "you" and "I" are. In the skprompt.txt, "I" seems to be used heavily for the AI persona. Is it the same here ("you" is the caller and "I" is the AI)?

Reply:
Made up. Yeah, I changed the wording in the prompt. Will update this sentence to match.